MTKA-1305: media repeatable

matthewguywright opened the pull request:
## Description

<!--
Include a summary of the change and some contextual information.
-->

<!--
Provide a link to the JIRA ticket (if any) for issue tracking purposes
-->

https://wpengine.atlassian.net/browse/

## Testing

<!--
Describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Also list any relevant details for your test configuration such as how to test the changes locally or in staging.
-->

## Screenshots

<!--
If this is a visual change include relevant screenshots about the behavior of the application before and after this change.
-->

## Documentation Changes

<!--
List corresponding changes to the documentation and updated wiki pages.
-->

## Dependant PRs

<!--
List any dependent PR's that are awaiting review. And in order to have these dependencies listed as part of the checks section, use the following syntax:

Depends on #1234
-->

matthewguywright committed:
added settings make repeatable checkbox

matthewguywright committed:
Update MediaFields.jsx

matthewguywright committed:
added MediaItems for graphql types

anthonyburchell commented on includes/settings/js/src/components/fields/MediaFields.jsx:
```diff
@@ -0,0 +1,53 @@
+	const [showRepeatable, setShowRepeatable] = useState(
+		data?.isRepeatable === true
+	);
+	const [showTitle, setShowTitle] = useState(data?.isTitle === true);
```
we wont need any concept of title for this one

anthonyburchell commented on includes/settings/js/src/components/fields/MediaFields.jsx:
```diff
@@ -0,0 +1,53 @@
+						ref={register}
+						defaultChecked={showRepeatable}
+						onChange={() => setShowRepeatable(!showRepeatable)}
+						disabled={editing || showTitle}
```
here also you are checking for showtitle and editing

matthewguywright committed:
Update MediaFields.jsx

anthonyburchell commented on includes/settings/js/src/components/fields/MediaFields.jsx:
```diff
@@ -0,0 +1,43 @@
+						type="checkbox"
+						id={`is-repeatable-${fieldId}`}
+						ref={register}
+						defaultChecked={showRepeatable}
```
try changing this to `value={showRepeatable}` instead of defaultChecked

matthewguywright committed:
Merge branch 'main' into feature/mw-mtka-1305-media

matthewguywright committed:
boilerplate

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
working media selector for multi

matthewguywright committed:
placeholders for buttons in rows

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
working select multi media

matthewguywright committed:
working preselect

matthewguywright committed:
add property updated on media selection

matthewguywright committed:
working delete on images

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
new layout

matthewguywright committed:
Merge branch 'main' into feature/mw-mtka-1305-media

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
updated per PR

matthewguywright committed:
Update MediaFields.jsx

matthewguywright committed:
added context to isRepeatable

matthewguywright committed:
dom errors fixed

matthewguywright committed:
media icon

mindctrl commented on includes/publisher/js/src/components/MediaUploader.jsx:
```diff
@@ -140,6 +193,283 @@ export default function MediaUploader({ modelSlug, field, required }) {
+		media.open();
+	}
+
+	if (field.isRepeatable && typeof field.isRepeatable === "boolean") {
```
```suggestion
	if (field.isRepeatableMedia && typeof field.isRepeatableMedia === "boolean") {
```

mindctrl replied:
@matthewguywright this `typeof field.isRepeatableMedia === "boolean"` check is what was breaking multi-select for me. The value of `field.isRepeatableMedia` is a string "true" instead of boolean true. Note that other properties use boolean values, so we should update this one to match.

![image](https://user-images.githubusercontent.com/1254870/157487168-a354fd03-ca35-471b-8dfe-264eed61ebac.png)

mindctrl commented on includes/settings/js/src/components/fields/MediaFields.jsx:
```diff
@@ -0,0 +1,43 @@
+
+const MediaFields = ({ register, data, editing, fieldId }) => {
+	const [showRepeatable, setShowRepeatable] = useState(
+		data?.isRepeatable === true
```
```suggestion
		data?.isRepeatableMedia === true
```

mindctrl commented on includes/publisher/js/src/components/MediaUploader.jsx:
```diff
@@ -140,6 +193,283 @@ export default function MediaUploader({ modelSlug, field, required }) {
+															<div
+																className={`value[${index}].remove-container p-2 me-sm-1`}
+															>
+																<button
```
Should this be the TertiaryButton component instead of manually doing the button?

mindctrl commented on includes/publisher/scss/_media-field.scss:
```diff
@@ -0,0 +1,139 @@
+@import "../../../includes/shared-assets/scss/_variables";
+
+.media {
```
Is there a reason we're doing scss files instead of Emotion here?

matthewguywright replied:
That's the pattern that's in place. It needs to be extracted from the scss and put into the repeatable fields if possible at some point.

mindctrl commented on includes/publisher/js/src/components/MediaUploader.jsx:
```diff
@@ -3,12 +3,23 @@ import { jsx, css } from "@emotion/react";
 import { __ } from "@wordpress/i18n";
+import TrashIcon from "../../../../components/icons/TrashIcon";
+import MediaIcon from "../../../../components/icons/MediaIcon";
+
```
We can use the `acm-icons` alias when importing icons, like we do in other files. However, the TrashIcon is not imported and returned in the [`index.js` file](https://github.com/wpengine/atlas-content-modeler/blob/main/includes/components/icons/index.js), like all the others, so that one would need to be updated to work like the rest.

mindctrl commented on includes/publisher/js/src/components/MediaUploader.jsx:
```diff
@@ -3,12 +3,23 @@ import { jsx, css } from "@emotion/react";
+	modelSlug,
+	field,
+	required,
+	errors,
```
```suggestion
```

`errors` isn't being used.

mindctrl commented on includes/publisher/js/src/components/MediaUploader.jsx:
```diff
@@ -3,12 +3,23 @@ import { jsx, css } from "@emotion/react";
+	modelSlug,
+	field,
+	required,
+	errors,
+	defaultError,
```
```suggestion
```
```suggestion
```

`errors` and `defaultError` aren't being used.

mindctrl commented on includes/publisher/js/src/components/MediaUploader.jsx:
```diff
@@ -19,30 +30,72 @@ export default function MediaUploader({ modelSlug, field, required }) {
+	 */
+	function setMultiMediaUrls() {
+		if (Array.isArray(value)) {
+			const imgArr = [];
```
```suggestion
```

Not used

mindctrl commented on includes/publisher/js/src/components/MediaUploader.jsx:
```diff
@@ -140,6 +193,283 @@ export default function MediaUploader({ modelSlug, field, required }) {
+														>
+															<div
+																className="me-lg-1 repeater-input-container"
+																name="repeaters"
```
```suggestion
```
`name` is an invalid attribute on a `div`

matthewguywright committed:
updated icons

matthewguywright committed:
Update _media-field.scss

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Revert "Update MediaUploader.jsx"

This reverts commit facb408fae2370ca70158a17576db449822ff7c3.

matthewguywright committed:
file type switching for image/icon

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
padding update for title

matthewguywright committed:
full working version

matthewguywright committed:
updated per PR

matthewguywright committed:
removed unused code

mindctrl commented on includes/publisher/js/src/components/MediaUploader.jsx:
```diff
@@ -3,46 +3,112 @@ import { jsx, css } from "@emotion/react";
 	// local
 	const imageRegex = /\.(gif|jpe?g|tiff?|png|webp|bmp)$/i;
+	const audioRegex = /\.(mp3|ogg|wav|m4a)$/i;
+	const fileRegex = /\.(pdf|doc?x|ppt?x|pps?x|odt|sls?x|psd)$/i;
```
Where did this list of file extensions come from? What about .txt files? The way this is written, any file type that doesn't match something results in no icon. 

Screenshot
![image](https://user-images.githubusercontent.com/1254870/157731053-e62de9aa-4dab-443d-86bf-a206058c7b48.png)

What if we refactored it so that if there's no match for a specific type, it defaults to the file icon?

matthewguywright replied:
That works. The list comes from the WP allowable upload types for the media modal.

matthewguywright replied:
Txt files are not allowed by WP by default.

mindctrl commented:
Just noting we also need some tests here.

matthewguywright committed:
added default file image per PR

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
Update CreateContentModelMediaFieldCest.php

matthewguywright committed:
tests

Sameedahmad commented:
media repeatable is really needed, Thank you @matthewguywright  for making it happen.

matthewguywright committed:
Merge branch 'main' into feature/mw-mtka-1305-media

matthewguywright committed:
icon updates

matthewguywright committed:
Update MediaUploader.jsx

matthewguywright committed:
updated to maintain checked is repeatable state for media items

matthewguywright committed:
Merge branch 'main' into feature/mw-mtka-1305-media

matthewguywright committed:
Updated post merge

nickcernis commented:
@matthewguywright If you create a repeating media field with “use as featured image” ticked and save it on the publisher screen, do you see the error below (“There was an error updating the Array field data”)?

The message seems to be coming from the `if ( 'media' === $field_type` logic starting line 399 of `includes/publisher/class-publisher-form-editing-experience.php`. We probably need to adapt it to account for the fact that the posted values can now be an array of thumbnail IDs and not only a single ID. (Maybe use the first thumbnail ID as the featured image if there's an array of thumbnail IDs?)

<img width="989" alt="Screenshot 2022-03-21 at 15 54 58" src="https://user-images.githubusercontent.com/647669/159288182-ad6ba086-ccc5-4cba-bb7a-e5b15d23bdd5.png">

andrewbotz commented on includes/components/icons/index.js:
```diff
@@ -65,6 +68,14 @@ export default function Icon({ type, size, noCircle }) {
+		case "multimedia":
+			return <ImgIcon />;
+		case "trash":
+			return <TrashIcon />;
```
Trash Icon exists on line 67, should be able to remove one or the other.

mindctrl commented:
Noting conversation from Slack, when the repeating option is enabled, it does not disable/prevent the enabling of the featured image option — and vice versa.

matthewguywright committed:
Update index.js

matthewguywright committed:
featured vs repeatable disabling

matthewguywright committed:
updated watch on dependent variables

nickcernis commented on includes/settings/js/src/components/fields/Form.jsx:
```diff
@@ -563,6 +566,7 @@ function Form({ id, position, type, editing, storedData, hasDirtyField }) {
 									defaultChecked={
 										storedData?.required === true
 									}
+									disabled={!!isRepeatableMedia}
```
It looks like the parent div of this input field needs to change the class if `isRepeatableMedia` is true, so that the label gets ghosted when the checkbox is disabled:

```
							<div
								className={
									!!isRepeatableMedia && "read-only editing"
								}
							>
```

nickcernis replied:
It looks like the parent div of this input field needs to change the class if `isRepeatableMedia` is true, so that the label gets ghosted when the checkbox is disabled:

```
							<div
								className={
									!!isRepeatableMedia && "read-only editing"
								}
							>
```

nickcernis commented on includes/settings/js/src/components/fields/MediaFields.jsx:
```diff
@@ -0,0 +1,49 @@
+	return (
+		<>
+			{data && (
+				<div className={"field"}>
```
Same here — the class should change when the field is not editable so that the label is dimmed.

```suggestion
				<div
					className={
						isFeaturedWatcher ? "field read-only editing" : "field"
					}
				>
```

nickcernis replied:
Same here — the class should change when the field is not editable so that the label is dimmed.

```suggestion
				<div
					className={
						isFeaturedWatcher ? "field read-only editing" : "field"
					}
				>
```

nickcernis commented on includes/settings/js/src/components/fields/MediaFields.jsx:
```diff
@@ -0,0 +1,49 @@
+						onChange={() =>
+							setShowRepeatableMedia(!showRepeatableMedia)
+						}
+						disabled={isFeaturedWatcher}
```
We should probably disable this when it's being edited to be consistent with the other “make repeatable” fields. (You can't change a field's repeatable status once it has been created.)

```suggestion
						disabled={editing || isFeaturedWatcher}
```

nickcernis commented on tests/acceptance/CreateContentModelMediaFieldCest.php:
```diff
@@ -8,6 +8,61 @@ public function _before( \AcceptanceTester $i ) {
+		$i->amOnPage( '/wp-admin/edit.php?post_type=candy' );
+		$i->click( 'Add New', '.wrap' );
+		$i->wait( 1 );
+		$i->seeElement( 'input[name="atlas-content-modeler[candy][positionxyz][0]"]:invalid' );
```
I think this test is failing because no field with a name of `atlas-content-modeler[candy][positionxyz][0]` exists until media has been selected from the media picker. (We only render the repeatable fields once a user has chosen media.)

If you apply the suggestion from Slack to render a hidden required input field when the user has not selected any media, you could then adapt this test to check for that field (by dropping the `[0]`):

```suggestion
		$i->seeElement( 'input[name="atlas-content-modeler[candy][positionxyz]"]:invalid' );
```

nickcernis commented on tests/acceptance/CreateContentModelMediaFieldCest.php:
```diff
@@ -8,6 +8,61 @@ public function _before( \AcceptanceTester $i ) {
+		$i->wait( 1 );
+
+		// Check that the first input in the index of available list items is rendering indicating a multiple being set.
+		$i->seeElement(
+			'input',
+			[ 'name' => 'atlas-content-modeler[candy][positionxyz][0]' ]
+		);
```
Same reason for the test failure here, I think: the element you're checking for doesn't exist until media has been selected.

You could check for something that is present instead? For example, instead of `seeElement()`:

```
$i->see( 'Manage Media', 'button[data-testid="media-uploader-manage-media-button"]' );
```

matthewguywright committed:
updated per PR

matthewguywright committed:
Update includes/settings/js/src/components/fields/MediaFields.jsx

Co-authored-by: Nick Cernis <nick@cern.is>

matthewguywright committed:
Update includes/settings/js/src/components/fields/MediaFields.jsx

Co-authored-by: Nick Cernis <nick@cern.is>

matthewguywright committed:
Update tests/acceptance/CreateContentModelMediaFieldCest.php

Co-authored-by: Nick Cernis <nick@cern.is>

matthewguywright committed:
updated test to look for something that is visible without needing to interact

matthewguywright committed:
Merge branch 'feature/mw-mtka-1305-media' of https://github.com/wpengine/atlas-content-modeler into feature/mw-mtka-1305-media

matthewguywright committed:
Update CreateContentModelMediaFieldCest.php

matthewguywright committed:
updated for required input

nickcernis commented on tests/acceptance/CreateContentModelMediaFieldCest.php:
```diff
@@ -8,6 +8,58 @@ public function _before( \AcceptanceTester $i ) {
+		$i->click( 'Add New', '.wrap' );
+		$i->wait( 1 );
+
+		// Check that the first input in the index of available list items is rendering indicating a multiple being set.
```
@matthewguywright This comment no longer seems to apply with the change to the line below. Should we delete it?

matthewguywright committed:
Update tests/acceptance/CreateContentModelMediaFieldCest.php

Co-authored-by: Nick Cernis <nick@cern.is>

mindctrl commented on includes/publisher/js/src/components/MediaUploader.jsx:
```diff
@@ -140,6 +211,336 @@ export default function MediaUploader({ modelSlug, field, required }) {
+					>
+						{field.name}
+					</label>
+					{field?.required && <p className="required">*Required</p>}
```
The "Required" text should be translatable. Something like this (eyeballing this):
```suggestion
					{field?.required && <p className="required">*{__("Required", "atlas-content-modeler")}</p>}
```

matthewguywright committed:
Update includes/publisher/js/src/components/MediaUploader.jsx

Co-authored-by: John Parris <john.parris@wpengine.com>

matthewguywright committed:
updated for linting

matthewguywright committed:
Update CreateContentModelMediaFieldCest.php

mindctrl reviewed (changes requested):
GraphQL queries on repeating media are causing WPGraphQL to freeze. The resolver needs an update, around line 541 or so in `custom-post-types-registration.php`.

```
					if ( $field['type'] === [ 'list_of' => 'MediaItem' ] ) {
						return array_map(
							function( $v ) use ( $context ) {
								return DataSource::resolve_post_object( (int) $v, $context );
							},
							$value
						);
					}
```

nickcernis committed:
fix: update GQL resolver for repeatable media

So that it returns an array of media objects if an array
of media IDs is provided as the value.
